Consolidate build and publish scripts

eliot-akira opened the pull request:
Resolves #15. The NPM scripts were getting long and unwieldy to edit inside `package.json`, so I moved them into separate files. I made a couple of other small improvements along the way.

adamziel commented on package.json:
```diff
@@ -13,19 +13,15 @@
     "dev": "npm run dev:web",
-    "build:php:web": "cd wasm-build/php && bash webworker-build-wasm.sh",
-    "build:php:node": "cd wasm-build/php && bash node-build-wasm.sh",
+    "build:php:web": "cd wasm-build/php && bash webworker-build-wasm.sh && bash web-publish.sh",
+    "build:php:node": "cd wasm-build/php && bash node-build-wasm.sh && bash node-publish.sh",
```
Since the npm script is called `build`, let's call the files consistently: `web-build.sh` and `node-build.sh`.

adamziel replied:
After some more thought I like your idea better – these scripts do publish the files after all.

adamziel commented on package.json:
```diff
@@ -13,19 +13,15 @@
+    "build:php:node": "cd wasm-build/php && bash node-build-wasm.sh && bash node-publish.sh",
     "build:php": "npm-run-all build:php:*",
-    "build:wp": "cd wasm-build/wordpress-data && bash prepare-wordpress.sh && bash bundle-data.sh",
-    "build:app:web": "esbuild --bundle src/web/app.mjs --outfile=dist-web/app.js --watch --external:xmlhttprequest",
```
Good catch!

eliot-akira replied:
Oh this reminds me: for my local dev branch I've added a few more tasks to `build:app:web` so that it does the same things as:

- `dev:web:app` (exists)
- `dev:web:web-worker` (without watch option)
- `dev:web:service-worker` (without watch option)
- `dev:web:html`

The last three `dev` tasks didn't have a corresponding `build` task.

It looks like:

```json
"build:app:web": "esbuild --bundle src/web/app.mjs --outfile=dist-web/app.js --external:xmlhttprequest && esbuild --bundle src/web/web-worker.js --outfile=dist-web/web-worker.js --external:xmlhttprequest && esbuild --bundle src/web/service-worker.js --outfile=dist-web/service-worker.js --external:xmlhttprequest && cp src/web/*.html dist-web/"
```

Maybe later it can be split into individual tasks and ran in parallel using `npm-run-all`.

---

Another idea: since there are tasks named `dev:web` and `dev:node`, it might make sense to name the build tasks `build:web` and `build:node`.

adamziel replied:
Oh the build tasks for the web-worker and service-worker would be really useful here – would you mind adding them to this PR? 

> Another idea: since there are tasks named dev:web and dev:node, it might make sense to name the build tasks build:web and build:node.

👍

adamziel commented on wasm-build/wordpress-data/prepare-wordpress.sh:
```diff
@@ -1,16 +1,18 @@
-tar -xzf wordpress-6.0.1.tar.gz
-rm wordpress-6.0.1.tar.gz
+# Download latest version of WordPress
+wp_tarfile=wordpress-latest.tar.gz
```
I feel pretty confident about defaulting to the latest minor release, e.g. `6.0.2`. Auto-upgrading to the next major release like `6.1.0` could break the build, though, if anything essential changes. I'd rather keep doing this manually every few months until there are more checks and balances in place.

eliot-akira replied:
Makes sense - that would be more stable and reliable than always pulling the newest version automatically. OK, updated.

adamziel commented on wasm-build/wordpress-data/web-publish.sh:
```diff
@@ -0,0 +1,14 @@
+dest_dir=$root_dir/dist-web
+
+cp ./docker-output/* $dest_dir/
+cp ./preload/wp-lazy-files.js $dest_dir/
```
I think the paths are relative to the current working directory, which means the `cp` here wouldn't find `./preload/wp-lazy-files.js` since it's executed from the root repo while `./prelload` is a subdirectory of `wasm-build/php`.

adamziel replied:
Wait, I'm wrong. This script is executed by npm with the correct cwd.

adamziel commented:
Great PR @eliot-akira, the code reads well! I'll also test it soon. Debugging https://github.com/WordPress/wordpress-wasm/issues/1 made my local setup quite unwieldy, but I will give this PR some testing soon.

eliot-akira commented:
Thank you for the review!

I'll make another commit so the WordPress download is pinned to a specific version.

In that case, what do you think about "caching" the tar file (by not removing it every time)?

```sh
wp_tarfile=wordpress-6.0.2.tar.gz
if [ ! -f $wp_tarfile ]; then
  wget https://wordpress.org/$wp_tarfile
fi
tar -xzf $wp_tarfile
```

The file could be added to `.gitignore`.


---

As an aside - I've continued trying to solve #24 with a "hand shake" between app.mjs and service worker (event message and response, then set iframe src) - but ran into difficulty with some fetch events from the iframe that don't have client ID for some reason, that cannot be distinguished from fetch events from the host. For this I'm testing on Chromium, and have observed #1 on occasion, "Aw snap".

adamziel reviewed (approved):
Great contribution as always @eliot-akira! 👍 I'm not merging yet as there is an ongoing discussion, but I'm approving already.

eliot-akira commented:
Thanks! This last change (mainly commit [6733446](https://github.com/WordPress/wordpress-wasm/pull/25/commits/6733446ecce9fef65fc2b860fe3d666e793c79e6)) turned out to be bigger than I expected. It'd be great if you could test drive the scripts to check that they all do everything as expected, same as before.

I reorganized the `dev` and `build` scripts in `package.json`, and renamed some of them for consistency. I also changed some of the `dev` scripts to run the corresponding `build` script with `--watch` option, to reduce duplicated commands and options.

- dev - Alias for dev:web
  - dev:web - Runs dev:web:\* in parallel
    - dev:web:app
    - dev:web:html
    - dev:web:serve
    - dev:web:service-worker
    - dev:web:web-worker
  - dev:node
- build - Alias for build:web
  - build:web - Runs build:web:\* in parallel
    - build:web:app
    - build:web:html
    - build:web:service-worker
    - build:web:web-worker
  - build:node
  - build:php - Runs build:php:\* sequentially
    - build:php:web
    - build:php:node
  - build:wp
  - build:app - Runs build:web and build:node sequentially
- clean - Runs clean:\* in parallel
  - clean:php
  - clean:wp

adamziel commented on package.json:
```diff
@@ -4,29 +4,31 @@
-    "build:app:web": "esbuild --bundle src/web/app.mjs --outfile=dist-web/app.js --watch --external:xmlhttprequest",
-    "build:app:node": "rm -rf dist-node/* && esbuild --bundle src/node/index.mjs --outfile=dist-node/index.mjs --platform=node --external:express --external:yargs --external:body-parser --external:cookie-parser --external:./src/node/node-php.js --target=node14.19.0 --format=esm && perl -pi.bak -e 's/\\.\\.\\/shared\\///g' dist-node/index.mjs && perl -pi.bak -e 's/\\.\\.\\/src\\/node\\//.\\//g' di ...
-    "build:app": "npm-run-all build:app:*",
+    "build:php:web": "cd wasm-build/php && bash webworker-build-wasm.sh && bash web-publish.sh",
```
```suggestion
    "build:php:web": "cd wasm-build/php && bash web-build-wasm.sh && bash web-publish.sh",
```

adamziel commented on wasm-build/php/web-publish.sh:
```diff
@@ -0,0 +1,9 @@
+root_dir=../..
+dest_dir=$root_dir/dist-web
+
+cp ./docker-output/webworker-* $dest_dir/
```
```suggestion
cp ./docker-output/php* $dest_dir/
```

adamziel commented:
It all works in my testing. There were a few merge conflicts with #28 so I went ahead and rebased this PR. Splendid work @eliot-akira, the scripts structure feels much cleaner now. Thank you so much!
